source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+        Queue, UR_QUEUE_INFO_CONTEXT, sizeof(ur_context_handle_t), &Context,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Context;
+}
+
+ur_device_handle_t getDevice(ur_queue_handle_t Queue) {
+    ur_device_handle_t Device;
+    auto Result = context.urDdiTable.Queue.pfnGetInfo(
+        Queue, UR_QUEUE_INFO_DEVICE, sizeof(ur_device_handle_t), &Device,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Device;
+}
+
+ur_program_handle_t getProgram(ur_kernel_handle_t Kernel) {
+    ur_program_handle_t Program;
+    auto Result = context.urDdiTable.Kernel.pfnGetInfo(
+        Kernel, UR_KERNEL_INFO_PROGRAM, sizeof(ur_program_handle_t), &Program,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Program;
+}
+
+} // namespace
+
+ur_result_t SanitizerInterceptor::allocateMemory(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+    const ur_usm_desc_t *Properties, ur_usm_pool_handle_t Pool, size_t Size,
+    void **ResultPtr, USMMemoryType Type) {
+    auto Alignment = Properties->align;
+    assert(Alignment == 0 || IsPowerOfTwo(Alignment));
+
+    auto ContextInfo = getContextInfo(Context);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+
+    if (Alignment == 0) {
+        Alignment = DeviceInfo->Alignment;
+    }
+
+    // Copy from LLVM compiler-rt/lib/asan
+    uptr RZLog = ComputeRZLog(Size);
+    uptr RZSize = RZLog2Size(RZLog);
+    uptr RoundedSize = RoundUpTo(Size, Alignment);
+    uptr NeededSize = RoundedSize + RZSize * 2;
+
+    void *Allocated = nullptr;
+
+    if (Type == USMMemoryType::DEVICE) {
+        UR_CALL(context.urDdiTable.USM.pfnDeviceAlloc(
+            Context, Device, Properties, Pool, NeededSize, &Allocated));
+    } else if (Type == USMMemoryType::HOST) {
+        UR_CALL(context.urDdiTable.USM.pfnHostAlloc(Context, Properties, Pool,
+                                                    NeededSize, &Allocated));
+    } else if (Type == USMMemoryType::SHARE) {
+        UR_CALL(context.urDdiTable.USM.pfnSharedAlloc(
+            Context, Device, Properties, Pool, NeededSize, &Allocated));
+    } else {
+        die("SanitizerInterceptor: unsupport memory type");
+    }
+
+    // Copy from LLVM compiler-rt/lib/asan
+    uptr AllocBegin = reinterpret_cast<uptr>(Allocated);
+    uptr AllocEnd = AllocBegin + NeededSize;
+    uptr UserBegin = AllocBegin + RZSize;
+    if (!IsAligned(UserBegin, Alignment)) {
+        UserBegin = RoundUpTo(UserBegin, Alignment);
+    }
+    uptr UserEnd = UserBegin + Size;
+    assert(UserEnd <= AllocEnd);
+
+    *ResultPtr = reinterpret_cast<void *>(UserBegin);
+
+    auto AllocInfo = std::make_shared<USMAllocInfo>(
+        USMAllocInfo{AllocBegin, UserBegin, UserEnd, NeededSize, Type});
+
+    // For updating shadow memory
+    {
+        std::scoped_lock<ur_shared_mutex> Guard(DeviceInfo->Mutex);
+        DeviceInfo->AllocInfos.emplace_back(AllocInfo);
```

Comment:
I don't think `AllocInfos` is ever removed from (apart from the clear on kernel launch). It's not impossible for software to do some sort of allocate/deallocate cycle prior to running any kernel, causing duplicate entries in `AllocInfos`. This might be especially problematic for host allocations where this is never cleared.

Reply:
Yes, you're right.
As what I mentioned in the above comments, currently I haven't completed the releasing logic.
In fact, the releasing logic needs much more consideration, so I'll leave it in another patch.